### PR TITLE
[Feature] Support  `iscrowd` attribute training in YOLOv5 

### PR DESCRIPTION
**Pull request by triple-mu**
This pr is aimed at  https://github.com/ultralytics/yolov5/issues/2720.

I verified the correctness of the code by visualization, but the mAP performance is low

The red box marks the ignore area, blue indicates the matching positive sample anchor, and cyan indicates the ignored positive sample anchor

image

### CODE REVIEW

**Comment by hhaAndroid** [comment]
@triple-Mu please rebase and resolve confict

**Comment by hhaAndroid** [comment]
@triple-Mu The commit record is still incorrect, so I guess there is a problem with the command

File: mmyolo/datasets/transforms/transforms.py [hunk 1]
```diff
@@ -352,6 +352,15 @@ class LoadAnnotations(MMDET_LoadAnnotations):
     time being, in order to speed up the pipeline, it can be excluded in
     advance."""
 
+    def __init__(
```
Comment (hhaAndroid) [inline mmyolo/datasets/transforms/transforms.py:355]: 不需要重写这个类，配置里面直接使用 mmdet 的 loadann 类就可以了

File: configs/yolov5/yolov5_s-with_ignore.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
+_base_ = '../_base_/default_runtime.py'
```
Comment (hhaAndroid) [inline configs/yolov5/yolov5_s-with_ignore.py:1]: 配置文件名要改

File: mmyolo/models/dense_heads/yolov5_head.py [hunk 3]
```diff
@@ -485,12 +485,35 @@ def loss_by_feat(
         # 1. Convert gt to norm format
         batch_targets_normed = self._convert_gt_to_norm_format(
             batch_gt_instances, batch_img_metas)
+        with_ignore = self.train_cfg.get(
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/yolov5_head.py:488]: 有在本地快速验证下正确性吗？ 可以抽取 coco数据子集，然后可视化分布，快速验证下

**Comment by triple-Mu** [comment]
> 

coco 跑了 170 轮

```shell
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.333                                                                                                        
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.520                                             
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.359                             
Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.188                                                                                                        
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.377                                                              
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.419                                                                                                        
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.283                                             
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.479                                                                                                        
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.533                                             
Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.334                                                                                                        
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.592                                             
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.666                                                                                                        
11/26 09:54:39 - mmengine - INFO - bbox_mAP_copypaste: 0.333 0.520 0.359 0.188 0.377 0.419
```

File: mmyolo/models/dense_heads/yolov5_head.py [hunk 4]
```diff
@@ -441,14 +446,25 @@ def loss(self, x: Tuple[Tensor], batch_data_samples: Union[list,
         """
 
         if isinstance(batch_data_samples, list):
-            losses = super().loss(x, batch_data_samples)
+            if self.ignore_iof_thr == -1.0:
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/yolov5_head.py:449]: 有点丑。要不不要改这个函数了，直接改 loss_by_feat

```python
        if self.ignore_iof_thr != -1:
            return self._loss_by_feat_with_ignore()
```

File: mmyolo/models/dense_heads/yolov5_head.py [hunk 5]
```diff
@@ -652,3 +690,199 @@ def _decode_bbox_to_xywh(self, bbox_pred, priors_base_sizes) -> Tensor:
...
+                    batch_size, self.num_base_priors, -1, h,
+                    w)[img_inds, priors_inds, :, grid_y_inds, grid_x_inds]
+
+                target_class = torch.full_like(pred_cls_scores, 0.)
+                target_class[range(batch_targets_scaled.shape[0]),
+                             class_inds] = 1.
+                loss_cls += self.loss_cls(
+                    pred_cls_scores,
+                    target_class,
+                    weight=not_ignore_weights[:, None].repeat(
```
Comment (BaofengZan) [inline mmyolo/models/dense_heads/yolov5_head.py:878]: 你好，现在是把ignore样本 无论是 box loss 还是 obj loss 还是 分类loss 都不计算了吗？  这样做和之前比有啥差异，有没有实验结果数据？

Comment (triple-Mu) [reply mmyolo/models/dense_heads/yolov5_head.py:878]: 上一次的实验已经跟coco对齐了，这次的是hha佬新改的 还需要继续实验

Comment (BaofengZan) [reply mmyolo/models/dense_heads/yolov5_head.py:878]: ok。那我直接理解最新的吧。/我看两个还是有些差异的。尤其在obj loss那里，之前的是把 ignore的gt赋值为1，没太理解。 现在直接用权重过滤了，没在gt上做文章，这里我还能理解。

Comment (hhaAndroid) [reply mmyolo/models/dense_heads/yolov5_head.py:879]: 对的

File: mmyolo/models/dense_heads/yolov5_head.py [hunk 6]
```diff
@@ -20,6 +21,15 @@
 from ..utils import make_divisible
 
 
+def get_prior_xy_info(index, num_base_priors, featmap_sizes):
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/yolov5_head.py:24]: typehint docstr

File: mmyolo/models/dense_heads/yolov5_head.py [hunk 7]
```diff
@@ -20,6 +21,15 @@
...
+    featmap_h, featmap_w = featmap_sizes
+    priors = index % num_base_priors
+    xy_index = index // num_base_priors
+    grid_y = xy_index // featmap_w
+    grid_x = xy_index % featmap_h
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/yolov5_head.py:29]: 这里写错了，应该 xy_index % featmap_w

File: mmyolo/models/dense_heads/yolov5_head.py [hunk 8]
```diff
@@ -484,6 +497,32 @@ def loss_by_feat(
         Returns:
             dict[str, Tensor]: A dictionary of losses.
         """
+        if self.ignore_iof_thr != -1:
+
+            # convert ignore gt
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/yolov5_head.py:502]: 加一个todo，因为我们还没有支持 fast 版本

File: mmyolo/models/dense_heads/yolov5_head.py [hunk 9]
```diff
@@ -652,3 +691,199 @@ def _decode_bbox_to_xywh(self, bbox_pred, priors_base_sizes) -> Tensor:
         pred_wh = (bbox_pred[:, 2:] * 2)**2 * priors_base_sizes
         decoded_bbox_pred = torch.cat((pred_xy, pred_wh), dim=-1)
         return decoded_bbox_pred
+
+    def _loss_by_feat_with_ignore(self, cls_scores: Sequence[Tensor],
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/yolov5_head.py:695]: 请加一下单元测试

**Comment by hhaAndroid** [comment]
@triple-Mu ci failed

**Comment by hhaAndroid** [comment]
@triple-Mu

**Review by hhaAndroid — APPROVED** [review]
